File: 2025/09 September/21 Max rectangle/main.py
# Link: https://www.geeksforgeeks.org/problems/max-rectangle/1


class Solution:
    def getMaxArea(self, heights):
        n = len(heights)
        s = []
        res = 0

        for i in range(n):
            while s and heights[s[-1]] >= heights[i]:
                tp = s.pop()
                width = i if not s else i - s[-1] - 1

                res = max(res, heights[tp] * width)
            s.append(i)

        while s:
            tp = s.pop()
            width = n if not s else n - s[-1] - 1
            res = max(res, heights[tp] * width)

        return res

    def maxArea(self, mat):
        n, m = len(mat), len(mat[0])
        heights = [0] * m
        ans = 0

        for i in range(n):
            for j in range(m):
                if mat[i][j] == 1:
                    heights[j] += 1
                else:
                    heights[j] = 0
            ans = max(ans, self.getMaxArea(heights))

        return ans


if __name__ == "__main__":
    obj = Solution()
    matrix = [[0, 1, 1, 0], [1, 1, 1, 1], [1, 1, 1, 1], [1, 1, 0, 0]]
    result = obj.maxArea(matrix)
    print(f"Maximum rectangle area is: {result}")

    matrix = [[0, 1, 1], [1, 1, 1], [0, 1, 1]]
    result = obj.maxArea(matrix)
    print(f"Maximum rectangle area is: {result}")

File: 2025/09 September/21 Max rectangle/main.cpp
// Link: https://www.geeksforgeeks.org/problems/max-rectangle/1

#include <iostream>
#include <vector>
#include <stack>
using namespace std;

class Solution
{
public:
    int getMaxArea(vector<int> &heights)
    {
        int n = heights.size();
        stack<int> s;
        int res = 0;

        for (int i = 0; i < n; i++)
        {
            while (!s.empty() && heights[s.top()] >= heights[i])
            {
                int tp = s.top();
                s.pop();

                int width = s.empty() ? i : i - s.top() - 1;

                res = max(res, heights[tp] * width);
            }
            s.push(i);
        }

        while (!s.empty())
        {
            int tp = s.top();
            s.pop();

            int width = s.empty() ? n : n - s.top() - 1;
            res = max(res, heights[tp] * width);
        }

        return res;
    }

    int maxArea(vector<vector<int>> &mat)
    {
        int n = mat.size(), m = mat[0].size();

        vector<int> heights(m, 0);
        int ans = 0;

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {

                if (mat[i][j] == 1)
                    heights[j]++;
                else
                    heights[j] = 0;
            }
            ans = max(ans, getMaxArea(heights));
        }

        return ans;
    }
};

int main()
{
    Solution obj;
    vector<vector<int>> mat = {{0, 1, 1, 0},
                               {1, 1, 1, 1},
                               {1, 1, 1, 1},
                               {1, 1, 0, 0}};
    int result = obj.maxArea(mat);
    cout << "Maximum rectangle area is: " << result << endl;

    mat = {{0, 1, 1},
           {1, 1, 1},
           {0, 1, 1}};
    result = obj.maxArea(mat);
    cout << "Maximum rectangle area is: " << result << endl;
    return 0;
}